### PR TITLE
fix: prevent PPO grid-socnav exit 139 on rerun (#5755)

### LINKED ISSUE #5755: bug: PPO grid-socnav extractor segfaults (exit 139) on rerun at main — blocks trace re-export
## Summary

Re-running PPO episodes at current `main` to capture per-step traces segfaults before any episode completes. A two-episode probe passed config/checkpoint/map preflight, loaded the exact release PPO checkpoint (SHA-256 `2b30df812bfcc737924b126b0763d69c567fe20716dc1c1eba8f56f926b49c1d`), then exited `139` (segmentation fault) inside `grid_socnav_extractor._infer_grid_output_dim`, producing zero episode rows.

## Impact

Blocks any exact-repeat or trace re-export of PPO episodes on `main`. The targeted trace re-export needed to render worked-example figures for the mined native candidates therefore has to run at the pinned execution commit `a307ef276d701f8d14dead1aa0513f44ee97c0b0` (where the 0.0.3 release episodes were originally executed) instead of `main`. That sidesteps the bug for that one job but does not fix it — anyone re-running PPO on `main` for reproducibility hits it.

## Repro

- On `main`, configure a PPO run over `classic_doorway_medium` (or `classic_realworld_double_bottleneck_high`), differential drive, horizon 600, with `record_simulation_step_trace: true`.
- Load the release PPO checkpoint (SHA-256 `2b30df81...`).
- Run: segmentation fault (exit 139) in `grid_socnav_extractor.py::_infer_grid_output_dim`; zero rows written.

## Likely area

`_infer_grid_output_dim` in the grid socnav feature extractor — a shape/dim inference path that appears to dereference an invalid tensor/observation-grid under the current `main` environment. Worth diffing that extractor between `a307ef276d701f8d14dead1aa0513f44ee97c0b0` and `main` (torch/numpy/observation-space drift is the prime suspect).

## Related

#5446 (candidate mining), #5498 (exact repeats after ORCA restore), #5580 (release provenance). Blocks the trace-re-export pipeline issue filed alongside this one.

### PR DESCRIPTION
**Pull request by ll7**
## Reviewer-critical summary

- **What changed:** `robot_sf.models` now applies the existing Torch 2.13/Python 3.12+ runtime guard before a caller can import Stable-Baselines3 for checkpoint loading. A subprocess regression test constructs a real CPU SB3 PPO optimizer across that import boundary.
- **Why now:** the public release `grid_socnav` PPO checkpoint reproduced exit 139 while `PPO.load` lazily entered Torch/Triton; importing the guard before SB3 made the same checkpoint load cleanly.
- **Claim boundary:** this establishes import/checkpoint-construction safety for the affected runtime path. It does not claim benchmark performance, trace completeness, or paper evidence.
- **Required validation:** focused PPO/extractor tests, the exact cached public checkpoint load, Ruff, and final `pr_ready_check` against `origin/main`.
- **Remaining blocker:** no code blocker remains for #5755; the full trace-capable benchmark rerun remains an intentionally excluded campaign act.

## Contract delta

- **New schema fields:** none.
- **New fail-closed blockers:** none; the existing version-scoped guard is reached earlier.
- **Compatibility impact:** only Torch 2.13.0 on Python 3.12+ preloads Triton; unaffected versions retain the existing no-op behavior. No observation, metric, or benchmark semantics change.

<details>
<summary>Repository governance and validation appendix</summary>

## Summary

Fixes the model-import ordering gap behind issue #5755's PPO `grid_socnav` exit-139 rerun failure.

## Linked Issues

- Closes #5755
- Issue: https://github.com/ll7/robot_sf_ll7/issues/5755

## Stack / Dependency

- Base dependency: `origin/main` (`a5ddc08e9` at branch creation)
- Required prior PRs: existing Torch/PPO runtime guard from #5556/#5724; no stacked dependency
- Safe to review independently: yes
- Fragmentation guard: no #5755 guardrail/checker/packet-refresh PR was found merged in the last 24 hours; this is the named runtime-fix progression slice.

## What Changed

- Added the runtime-guard call to `robot_sf/models/__init__.py`, immediately before direct SB3 consumers can load a resolved checkpoint.
- Added a subprocess regression test in `tests/test_seed_utils.py` that imports the model package, imports SB3, and constructs a CPU PPO optimizer.
- No configs, benchmark schemas, metric semantics, checkpoints, trace artifacts, or transient queue-routing state changed.

## Why It Matters

- Added value: prevents the affected PPO checkpoint path from entering the lazy Torch Dynamo/Triton import sequence after TensorFlow is initialized.
- Expected impact: the exact public checkpoint can be loaded on the current Python 3.13/Torch 2.13 CPU runtime, allowing the downstream trace rerun to proceed.
- Why now: the failure blocks #5756's trace re-export follow-up and was reproducible at current `main`.

## Research Result Guidance

- Target claim / hypothesis / blocker this should affect: remove the PPO checkpoint-construction blocker for #5755.
- Comparator or baseline, if applicable: pre-change current-main import order, which exited 139; post-change guarded import order, which loads successfully.
- Evidence tier: targeted smoke
- Result classification: blocker-resolution
- Decision or stop rule, if applicable: stop code work after exact checkpoint construction and focused tests pass; do not promote the result as benchmark evidence.
- Parent issue, claim map, registry, context note, or synthesis surface to update: #5755; no evidence registry update is warranted.
- New research/benchmark/metric/paper-facing analysis tool, if any: NA - support/runtime fix; no analysis tool or claim surface was added.

## Domain-Aware Approval

- Required for this PR: yes - a bounded domain review is recorded because the targeted smoke touches a benchmark execution path, while no metric or paper claim changes.
- Domains reviewed: runtime compatibility, benchmark execution safety
- Status: waived
- Approver/review source or waiver: waived for this support/runtime fix under the maintainer-value boundary; no benchmark interpretation or paper claim is introduced.
- Validity checklist:
  - Target claim/hypothesis: remove the PPO checkpoint-construction crash blocker; do not claim a benchmark result.
  - Comparator or split/evidence validity: pre-change current-main import order versus the guarded import order; no split or metric comparison is asserted.
  - Fallback/degraded exclusions: no fallback or degraded execution was counted as success; no episode campaign ran.
  - Claim boundary: CPU checkpoint-construction/import safety only; not benchmark evidence.
  - Implementation integrity vs experimental validity: implementation integrity is tested; experimental validity remains unassessed and out of scope.

## Falsification / Non-Transfer Check

- Did the mechanism activate? yes, in the affected Torch 2.13/Python 3.13 subprocess.
- Did the intervention change command source, selected command, trajectory, or route progress? NA - no episode was run.
- Did the scenario actually contain the targeted failure mode? yes for the public checkpoint-construction probe; the native pre-change failure was exit 139 in the Torch/Triton stack.
- Result route: continue to the separately authorized trace rerun after merge.
- Follow-up question or issue for weak, negative, or non-transfer results: whether the downstream trace rerun completes on its target host.

## Next Empirical Action

- Rerun needed: yes - the full trace-capable rerun is downstream work.
- Extractor or analysis tool needed: no - existing trace/export tooling is the consumer.
- Artifact missing or unavailable: no durable benchmark output was produced in this PR.
- Stop / revise / continue decision: continue after merge with the pinned CPU rerun on its designated host.
- Proposed child issue or existing follow-up: #5756.

## Validation / Proof

- `uv run pytest -q tests/test_seed_utils.py tests/test_grid_socnav_extractor.py tests/baselines/test_ppo_planner.py tests/training/test_ppo_policy.py tests/test_map_runner_ppo.py tests/benchmark/test_helper_catalog.py tests/validation/test_check_legacy_ppo_snapshot_parity.py` — 72 passed.
- `uv run ruff check robot_sf/models/__init__.py tests/test_seed_utils.py` — passed.
- `uv run ruff format --check robot_sf/models/__init__.py tests/test_seed_utils.py` — passed.
- `git diff --check origin/main...HEAD` — passed.
- `PYTHONFAULTHANDLER=1 timeout 180s uv run python <cached public PPO checkpoint load probe>` — exit 0; loaded `MultiInputActorCriticPolicy`. The equivalent pre-change probe exited 139 in `torch._compile`/Triton during SB3 optimizer construction.
- `PR_READY_MODE=final BASE_REF=origin/main scripts/dev/pr_ready_check.sh` — exit 0 with the PR body supplied to the body-aware gate; final readiness stamp passed on fresh merged HEAD `9d59973e` with a clean tree.
- No full benchmark campaign run, no Slurm/GPU submission, and no paper/dissertation claim edits.

## Risks / Rollout

- Compatibility risks: the guard imports Triton during `robot_sf.models` import only on the already-supported affected Torch/Python version pair; this adds startup work to model consumers on that pair.
- Failure modes: an incompatible or missing optional Torch/Triton installation still fails through the existing runtime/import path rather than producing benchmark fallback evidence.
- Rollback or fallback plan: revert this two-file commit; no benchmark fallback mode is introduced.

## Docs / Provenance

- Updated docs: none; this is an internal runtime ordering fix.
- Relevant design or provenance notes: `robot_sf/common/seed.py` owns the existing version-scoped guard.
- Any assumptions that need to be preserved: model resolution remains the supported boundary before direct PPO checkpoint loading; no transient host/packet routing state is tracked here.

## Downstream Propagation

- Parent issue updated: no - issue/PR comments are not authorized for this task.
- Claim map / benchmark report updated: no - no benchmark evidence was generated.
- Leaderboard / artifact catalog updated: no - no promoted artifact changed.
- Registry or config index updated: no - model registry data is unchanged.
- Context index / memory note updated: no - the PR is the durable implementation surface.
- Follow-up issue opened for deferred propagation: no - #5756 already tracks the downstream trace re-export.
- Not applicable because: this PR is a bounded runtime bug fix and the PR body is the sole authorized durable propagation surface.

## Follow-Up Issues

- Deferred work: run the pinned trace-capable rerun and review its outputs after this fix is merged.
- Issues opened for follow-up: none.

## Reviewer Notes

- Verify that `robot_sf.models` reaches `_configure_torch_213_runtime()` before a direct SB3 import and that the guard remains version-scoped.
- Verify the subprocess test exercises optimizer construction rather than only checking environment variables.
- This PR deliberately does not run the full benchmark/campaign or promote any research claim.
- Friction encountered: the required `gh issue view --comments` command is currently blocked by GitHub's Projects Classic GraphQL deprecation warning; the live issue was verified through the REST fallback. Existing issue/PR tracking for that friction was checked; no duplicate friction issue was opened.

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved startup compatibility for reinforcement-learning workflows using supported Torch and Stable-Baselines3 configurations.
  * Prevented a runtime initialization issue that could occur when loading models in certain Python environments.

* **Tests**
  * Added coverage to verify reliable model initialization across supported runtime configurations.

### CODE REVIEW

**Comment by ll7** [comment]
Gate findings for PR #5773 at head `9d59973e24e9725218bdc2a550697b9ea94c8ffc`:

Open review-thread enumeration: **4 open review threads**; three Gemini Code Assist threads and one CodeRabbit thread. All four are addressed in phase 2:

1. Gemini medium: importing `robot_sf.common.seed` already runs the Torch 2.13 guard, so calling the helper again from `robot_sf.models` performs redundant version/optional-module work — **fixed in phase 2** by consuming the module-level guard result.
2. Gemini medium: the duplicate helper call should be replaced by a reference to `_TORCH_213_RUNTIME_GUARD_APPLIED` — **fixed in phase 2** with the same single-import guard reference.
3. Gemini medium: the subprocess test must not require Triton when the optional package is unavailable — **fixed in phase 2** with conditional `find_spec`/`sys.modules` validation.
4. CodeRabbit minor/medium: the same optional-Triton assertion can fail on CPU-only installs — **fixed in phase 2** by the same conditional assertion.

No finding is waived or spun off. The PR remains a bounded Torch 2.13/Python 3.12+ checkpoint-construction fix; no benchmark result or fallback/degraded evidence is claimed.

**Comment by ll7** [comment]
Gate findings closure for PR #5773 at repaired head `77fa3a6fa17dbec9dfdd7932ba9fc4bdb95c6ef9`:

Open review-thread enumeration after phase 2: **0 open review threads**. The three Gemini threads and one CodeRabbit thread are resolved; no other CodeRabbit thread is open.

1. Gemini medium: importing `robot_sf.common.seed` already runs the Torch 2.13 guard, so the model package must not call it twice — **fixed** by consuming the module-level guard result.
2. Gemini medium: reference `_TORCH_213_RUNTIME_GUARD_APPLIED` to preserve the ordering without a second probe — **fixed** with the single-import guard reference.
3. Gemini medium: the subprocess test must tolerate optional Triton being absent — **fixed** with conditional `find_spec`/`sys.modules` validation.
4. CodeRabbit minor/medium: the same optional-Triton assertion could fail on CPU-only installs — **fixed** by the same conditional assertion.

No finding is waived or spun off. Focused tests, Ruff, format, and diff checks passed; the PR remains a bounded runtime compatibility fix with no benchmark claim.

**Comment by ll7** [comment]
Gate staleness delta for PR #5773: after PR #5757 merged, the unchanged reviewed Torch runtime repair was rebased onto new `main`. New head: `fe826c102` (full SHA is visible in the PR head); prior four-thread/0-thread verdict stands. CI is rerunning against the current base.

**Comment by ll7** [comment]
Gate findings delta for PR #5773 at head `fe826c1020e4aefddc6bacd1202ed701efc863df`:

Open review-thread enumeration: **0 open review threads**. The current patch is unchanged from the previously reviewed repair at `77fa3a6fa17dbec9dfdd7932ba9fc4bdb95c6ef9` after the rebase; no new substantive diff finding was identified.

Prior review findings remain individually fixed:
1. Gemini medium: avoid invoking the Torch runtime guard twice through `robot_sf.models` — **fixed** by consuming the module-level guard result.
2. Gemini medium: retain the guard ordering without a second version/optional-module probe — **fixed** with `_TORCH_213_RUNTIME_GUARD_APPLIED`.
3. Gemini medium: tolerate CPU-only environments without optional Triton — **fixed** with conditional availability/module validation.
4. CodeRabbit minor/medium: the same optional-Triton assertion could fail on CPU-only installs — **fixed** by the same conditional assertion.

No finding is waived or spun off. The current body still truthfully says `Closes #5755` for the runtime blocker and maps the excluded downstream trace rerun to existing follow-up #5756. The branch is stale against current `main` and must be refreshed before merge.

**Comment by ll7** [comment]
Gate staleness delta for PR #5773: the unchanged reviewed runtime repair was rebased onto current `main` because its CI base was stale. New head: `93e2b7c9d3e3614498a515fdec1c3cdfdfa6c315`. Post-push enumeration confirms **0 open review threads**; the prior four-finding/0-thread verdict stands and no new finding was introduced.

**Comment by ll7** [comment]
Gate staleness delta for PR #5773: after PR #5767 merged, the unchanged reviewed runtime repair was rebased onto the new `main` tip because its previous CI base became stale. New head: `e41ac732ce6526e1bd9cc3eb332c9194301a8d43`. Post-push enumeration confirms **0 open review threads**; the prior four-finding/0-thread verdict stands and no new finding was introduced. CI must complete for this refreshed head.

**Comment by ll7** [comment]
Gate staleness delta for PR #5773: `main` advanced again during the green CI run, so the unchanged reviewed runtime repair was rebased onto the new tip. New head: `1001d64eeeb74deb64c466e617accf1c3d9219bf`. Post-push enumeration confirms **0 open review threads**; the prior four-finding/0-thread verdict stands and no new finding was introduced. CI must complete for this refreshed head.

**Comment by ll7** [comment]
Gate staleness delta for PR #5773: PR #5778 merged while the previous green run was completing, so the unchanged reviewed runtime repair was rebased onto current `main`. New head: `614322a1837c12c297e533a53fed3a8c8d7e48a9`. Post-push enumeration confirms **0 open review threads**; the prior four-finding/0-thread verdict stands and no new finding was introduced. CI must complete for this refreshed head.

**Comment by ll7** [comment]
Gate freshness delta for PR #5773 at head `6bd2fd3d3557620126898a3cb761ec970f1270bb`:

- Open review-thread enumeration: **0 open review threads**.
- Prior gate review at `77fa3a6` stands; all four prior findings remain individually fixed, with no finding waived or spun off.
- The branch was rebased onto current `origin/main` solely for stale-CI freshness; no substantive PR change was introduced.
- The current body still says `Closes #5755`, identifies #5756 as the downstream trace rerun, and has no untracked blocker.

Fresh CI is now required for this head.

**Comment by ll7** [comment]
auto-promoted: complete + green, never un-drafted (draft-parking guard)
